### PR TITLE
refac: break listener into event listener and tx listener and manage txListener in server layer

### PR DESCRIPTION
**Pull request by himank**
This allows unifying the listeners and with this approach we just can have a single eventListener and that can be used by any pluggable TxListeners.

### CODE REVIEW

File: store/kv/listener.go [hunk 1]
```diff
@@ -16,48 +16,87 @@ package kv
...
 
-func GetListener(ctx context.Context) Listener {
-	l, ok := ctx.Value(ListenerCtxKey{}).(Listener)
-	if l != nil && ok {
-		return l
-	} else {
-		return &NoListener{}
+func GetEventListener(ctx context.Context) EventListener {
+	a := ctx.Value(EventListenerCtxKey{})
+	fmt.Printf("GetEventListener %T%v\n", a, a)
```
Comment (mattayres) [inline store/kv/listener.go:94]: Probably not meant to be left here?

File: server/services/v1/tx_listener.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	// OnPostCommit is called after committing the transaction. This means the transaction is complete and any post
+	// activity can happen as part of this implementation.
+	OnPostCommit(context.Context, kv.EventListener) error
+	// OnRollback is called when the user transaction is rolled back.
+	OnRollback(context.Context, kv.EventListener)
+}
+
+type NoopTxListener struct{}
+
+func (l *NoopTxListener) OnCommit(context.Context, transaction.Tx, kv.EventListener) error {
```
Comment (mattayres) [inline server/services/v1/tx_listener.go:37]: Maybe rename this to `OnPreCommit` now that there's an `OnPostCommit`.

File: store/kv/fdb.go [hunk 3]
```diff
@@ -306,6 +313,10 @@ func (b *fbatch) SetVersionstampedValue(_ context.Context, _ []byte, _ []byte) e
 	return fmt.Errorf("batch doesn't support setting versionstamped value")
 }
 
+func (b *fbatch) SetVersionstampedKey(_ context.Context, _ []byte, _ []byte) error {
+	return fmt.Errorf("batch doesn't support setting versionstamped key")
```
Comment (mattayres) [inline store/kv/fdb.go:317]: Should this return an error?

**Review by mattayres — APPROVED** [review]

File: store/kv/fdb.go (hunk 3 above)
Comment (himank) [reply store/kv/fdb.go:317]: realized after adding it in fdb layer that fdb is not returning an error for this. Makes sense, will remove it.

File: server/services/v1/tx_listener.go (hunk 2 above)
Comment (himank) [reply server/services/v1/tx_listener.go:37]: sg

File: store/kv/fdb.go (hunk 3 above)
Comment (efirs) [reply store/kv/fdb.go:317]: Maybe we can just remove batch in one of the next diffs, it's unused anyway.

**Review by efirs — APPROVED** [review]

**Pull request merged**